File: pharmacy-api/local-api/controllers/task_controller.go
package controllers

import (
	"encoding/json"
	"log"
	"net/http"
	"strconv"
	"time"

	controllers "pharmacy-api/shared/controllers"
	models "pharmacy-api/shared/models"

	"github.com/gin-gonic/gin"
	"github.com/go-redis/redis/v8"
	amqp "github.com/rabbitmq/amqp091-go"
)

func MakeTask(route string, task string, redisDB *redis.Client, broker *models.Broker) gin.HandlerFunc {
	return func(c *gin.Context) {

		ctx := c.Request.Context()

		var taskContext models.RequestContext

		var err error

		query := make(map[string]any)

		user, _ := c.Get("user")
		claims := user.(*models.Claims)

		if taskContext.Claims, err = json.Marshal(claims); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error": "Internal error",
			})
			log.Print(err.Error())
		}

		switch task {
		case "get":
			var idInt int
			idInt, _ = strconv.Atoi(c.Query("id"))
			if route != "users" {
				query["id"] = idInt
			} else {
				query["id"] = c.Query("id")
			}
			query["q"] = c.Query("q")
			query["page"] = c.Query("page")
			query["limit"] = c.Query("limit")

			if err != nil && c.Query("id") != "" {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "Invalid request body",
				})
				log.Print(err.Error())
				return
			}

			if taskContext.Query, err = json.Marshal(query); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{
					"error": "Internal error",
				})
				log.Print(err.Error())
			}
		case "post":
			if err := c.ShouldBindJSON(&taskContext.Context); err != nil {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "Invalid request body",
				})
				log.Print(err.Error())
				return
			}
		case "patch":
			query["id"], err = strconv.Atoi(c.Query("id"))
			if err != nil && c.Query("id") != "" {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "Invalid request body",
				})
				log.Print(err.Error())
				return
			}
			if taskContext.Query, err = json.Marshal(query); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{
					"error": "Internal error",
				})
				log.Print(err.Error())
			}

			if err := c.ShouldBindJSON(&taskContext.Context); err != nil {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "Invalid request body",
				})
				log.Print(err.Error())
				return
			}
		case "delete":
			query["id"], err = strconv.Atoi(c.Query("id"))
			if err != nil && c.Query("id") != "" {
				c.JSON(http.StatusBadRequest, gin.H{
					"error": "Invalid request body",
				})
				log.Print(err.Error())
				return
			}
			if taskContext.Query, err = json.Marshal(query); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{
					"error": "Internal error",
				})
				log.Print(err.Error())
			}
		}
		var taskContextJSON []byte
		if taskContextJSON, err = json.Marshal(taskContext); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error": "Internal error",
			})
			log.Print(err.Error())
		}
		taskID, err := controllers.RandomGen()
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error": "Internal error",
			})
			log.Print(err.Error())
			return
		}

		key := "local_task:" + taskID

		if err = redisDB.HSet(ctx, key, map[string]interface{}{
			"status":  "pending",
			"task":    task,
			"route":   route,
			"result":  "",
			"context": taskContextJSON,
		}).Err(); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error": "Internal error",
			})
			log.Print(err.Error())
			return
		}

		if err = redisDB.Expire(ctx, key, 5*time.Minute).Err(); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error": "Internal error",
			})
			log.Print(err.Error())
			return
		}

		err = broker.Channel.Publish(
			"local_exchange",
			route,
			false,
			false,
			amqp.Publishing{
				ContentType: "application/json",
				Body:        []byte(key),
				Timestamp:   time.Now(),
			},
		)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error": "Internal error",
			})
			log.Print(err.Error())
			return
		}

		c.JSON(http.StatusOK, gin.H{
			"TaskID": taskID,
		})
	}
}

File: pharmacy-api/src/local-api/controllers/task_controller.ts
import { Request, Response } from 'express';
import Redis from 'ioredis';
import { Channel } from 'amqplib';
import { randomBytes } from 'crypto';
import { RequestContext, Claims } from '@pharmacy/src/shared/models/models';

function randomGen(): string {
  return randomBytes(8).toString('hex');
}

declare global {
  namespace Express {
    interface Request {
      user?: Claims;
    }
  }
}

export function makeTask(route: string, task: string, redisDB: Redis, channel: Channel) {
  return async (req: Request, res: Response): Promise<void> => {
    let taskContext: RequestContext = {
      query: {},
      context: {},
      claims: {}
    };

    let query: Record<string, any> = {};

    try {
      const claims = req.user;
      
      if (claims) {
        taskContext.claims = claims;
      }

      switch (task) {
        case "get":
          const idParam = req.query.id as string;
          
          if (route !== "users") {
            const idInt = parseInt(idParam);
            if (isNaN(idInt) && idParam) {
              res.status(400).json({
                error: "Invalid request body",
              });
              return;
            }
            query["id"] = isNaN(idInt) ? undefined : idInt;
          } else {
            query["id"] = idParam || undefined;
          }
          
          query["q"] = req.query.q || "";
          query["page"] = req.query.page || "";
          query["limit"] = req.query.limit || "";
          
          taskContext.query = query;
          break;
          
        case "post":
          taskContext.context = req.body;
          break;
          
        case "patch":
          const patchId = parseInt(req.query.id as string);
          if (isNaN(patchId) && req.query.id) {
            res.status(400).json({
              error: "Invalid request body",
            });
            return;
          }
          query["id"] = isNaN(patchId) ? undefined : patchId;
          taskContext.query = query;
          taskContext.context = req.body;
          break;
          
        case "delete":
          const deleteId = parseInt(req.query.id as string);
          if (isNaN(deleteId) && req.query.id) {
            res.status(400).json({
              error: "Invalid request body",
            });
            return;
          }
          query["id"] = isNaN(deleteId) ? undefined : deleteId;
          taskContext.query = query;
          break;

        case "permissions":
          const id = req.query.id as string;
          if (id && isNaN(parseInt(id))) {
            res.status(400).json({
              error: "Invalid request body",
            });
            return;
          }
          
          query["id"] = id ? parseInt(id) : undefined;
          query["q"] = req.query.q || "";
          query["page"] = req.query.page || "";
          query["limit"] = req.query.limit || "";
          
          taskContext.query = query;
          break;
      }
      
      const taskID = randomGen();
      const key = `local_task:${taskID}`;
      
      await redisDB.hset(key, {
        status: "pending",
        task: task,
        route: route,
        result: "",
        context: JSON.stringify(taskContext),
      });
      
      await redisDB.expire(key, 300);
      
      const published = channel.publish(
        "local_exchange",
        route,
        Buffer.from(key),
        {
          contentType: "application/json",
          timestamp: Date.now(),
        }
      );
      
      if (!published) {
        console.error("Failed to publish message to RabbitMQ");
        res.status(500).json({
          error: "Internal error",
        });
        return;
      }
      
      res.status(200).json({
        TaskID: taskID,
      });
      
    } catch (err) {
      console.error("Error in makeTask:", err);
      res.status(500).json({
        error: "Internal error",
      });
    }
  };
}